### ISSUE TITLE
Optmized TWCS single-partition reader opens sstables unnecessarily

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#8432]
While analyzing #8411 I came up with a scenario which _should_ reproduce the problem reported there, but for some reason didn't. I analyzed why my reproducer didn't reproduce the problem and I found the reason: a different bug. Looking at `time_series_sstable_set::create_single_key_sstable_reader`, we see at the end the following `filter` passed to the sstable reader merger:
```cpp
    auto filter = [pk_filter = std::move(pk_filter), ck_filter = std::move(ck_filter), &stats]
        (const sstable& sst) {
            if (pk_filter(sst)) {
                return true;
            }

            ++stats.sstables_checked_by_clustering_filter;
            if (ck_filter(sst)) {
                ++stats.surviving_sstables_after_clustering_filter;
                return true;
            }

            return false;
    };

    return make_clustering_combined_reader(
            std::move(schema), std::move(permit), fwd_sm,
            make_min_position_reader_queue(std::move(create_reader), std::move(filter)));

```
This is incredibly stupid (I must have been very tired when implementing it...): note that the filter includes an sstable as soon as it passes through the `pk_filter` (which is the bloom filter combined with a token-position-range check), never looking at `ck_filter` (which uses the clustering-range metadata of the sstable). Furthermore, it sometimes includes sstables which didn't pass the `pk_filter` if they passed the `ck_filter`. Fortunately, this doesn't cause relevant data to be skipped (the filter always includes a superset of the set of sstables that contain the relevant data), but it causes sstables to be unnecessarily opened, resulting in a performance loss (these sstables would immediately return eof), and apparently masks other bugs, making them harder to reproduce...

**Comment by kbr-** [comment]
Note: before backporting the fix to 4.4 we should also fix #8411 (I already have a reproducer and @haaawk has a fix I think).

**Comment by avikivity** [comment]
Already backported to all vulnerable branches, removing "Backport candidate" label.
